go-blip data race detected in unit tests

From http://drone.couchbase.io/couchbase/sync_gateway/3900:

```

2018-01-15T23:47:51.156Z Sync+:     Sending 1 changes ...  (as ADMIN)
==================
WARNING: DATA RACE
Write at 0x00c4209fa008 by goroutine 8:
  github.com/couchbase/sync_gateway/rest.(*blipSyncContext).register.func1()
      /drone/godeps/src/github.com/couchbase/sync_gateway/rest/blip_sync.go:128 +0x91
  github.com/couchbase/go-blip.(*Context).dispatchRequest()
      /drone/godeps/src/github.com/couchbase/go-blip/context.go:145 +0x1dd
  github.com/couchbase/go-blip.(*receiver).getPendingRequest.func1()
      /drone/godeps/src/github.com/couchbase/go-blip/receiver.go:223 +0x8c
  github.com/couchbase/go-blip.(*Message).asyncRead.func1()
      /drone/godeps/src/github.com/couchbase/go-blip/message.go:326 +0x6b

Previous read at 0x00c4209fa008 by goroutine 203:
  github.com/couchbase/sync_gateway/rest.(*blipHandler).sendBatchOfChanges()
      /drone/godeps/src/github.com/couchbase/sync_gateway/rest/blip_sync.go:304 +0x37b
  github.com/couchbase/sync_gateway/rest.(*blipHandler).sendChanges.func2()
      /drone/godeps/src/github.com/couchbase/sync_gateway/rest/blip_sync.go:268 +0xb1
  github.com/couchbase/sync_gateway/rest.(*blipHandler).sendChanges.func3()
      /drone/godeps/src/github.com/couchbase/sync_gateway/rest/blip_sync.go:288 +0x7e1
  github.com/couchbase/sync_gateway/rest.generateContinuousChanges()
      /drone/godeps/src/github.com/couchbase/sync_gateway/rest/changes_api.go:633 +0x90f
  github.com/couchbase/sync_gateway/rest.(*blipHandler).sendChanges()
      /drone/godeps/src/github.com/couchbase/sync_gateway/rest/blip_sync.go:273 +0x64b

Goroutine 8 (running) created at:
  github.com/couchbase/go-blip.(*Message).asyncRead()
      /drone/godeps/src/github.com/couchbase/go-blip/message.go:324 +0x1a7
  github.com/couchbase/go-blip.(*receiver).getPendingRequest()
      /drone/godeps/src/github.com/couchbase/go-blip/receiver.go:222 +0x539
  github.com/couchbase/go-blip.(*receiver).processFrame()
      /drone/godeps/src/github.com/couchbase/go-blip/receiver.go:178 +0x560
  github.com/couchbase/go-blip.(*receiver).handleIncomingFrame()
      /drone/godeps/src/github.com/couchbase/go-blip/receiver.go:167 +0xaa9
  github.com/couchbase/go-blip.(*receiver).parseLoop()
      /drone/godeps/src/github.com/couchbase/go-blip/receiver.go:90 +0x153

Goroutine 203 (running) created at:
  github.com/couchbase/sync_gateway/rest.(*blipHandler).handleSubscribeToChanges()
      /drone/godeps/src/github.com/couchbase/sync_gateway/rest/blip_sync.go:237 +0x5cf
  github.com/couchbase/sync_gateway/rest.userBlipHandler.func1()
      /drone/godeps/src/github.com/couchbase/sync_gateway/rest/blip_sync.go:58 +0xbe
  github.com/couchbase/sync_gateway/rest.(*blipSyncContext).register.func1()
      /drone/godeps/src/github.com/couchbase/sync_gateway/rest/blip_sync.go:137 +0x3b6
  github.com/couchbase/go-blip.(*Context).dispatchRequest()
      /drone/godeps/src/github.com/couchbase/go-blip/context.go:145 +0x1dd
  github.com/couchbase/go-blip.(*receiver).getPendingRequest.func1()
      /drone/godeps/src/github.com/couchbase/go-blip/receiver.go:223 +0x8c
  github.com/couchbase/go-blip.(*Message).asyncRead.func1()
      /drone/godeps/src/github.com/couchbase/go-blip/message.go:326 +0x6b
==================
2018-01-15T23:47:51.156Z Sync: MSG#4~ "rev" ...  (as ADMIN)
2018-01-15T23:47:51.156Z Sync+: Inserting rev "foo-3" 1-abc history="", array = []string{"1-abc"} ...  (as ADMIN)
2018-01-15T23:47:51.157Z Sync+: MSG#4~    --> OK ...  (as ADMIN)
2018-01-15T23:47:51.157Z Changes+: Notifying that "test_data_bucket-a2a99769200dd8ee22cfa5d79a3e6492" changed (keys="{*}") count=5
2018-01-15T23:47:51.159Z BLIP: Push RPY#3
..........
2018-01-15 23:47:51.194556 I | got changes message: MSG#5~
2018-01-15 23:47:51.194600 I | changes body: [[4,"foo-4","1-abc"]], err: <nil>
2018-01-15T23:47:51.194Z BLIP: Push RPY#5
2018-01-15T23:47:51.194Z Changes+: changeListener.Stop() called
--- FAIL: TestContinousChangesSubscription (0.23s)
	testing.go:699: race detected during execution of test

```

--- a/rest/blip_api_test.go
+++ b/rest/blip_api_test.go
@@ -217,7 +217,7 @@ func TestContinousChangesSubscription(t *testing.T) {
 	subChangesResponse := subChangesRequest.Response()
 	assert.Equals(t, subChangesResponse.SerialNumber(), subChangesRequest.SerialNumber())
 
-	for i := 1; i < 5; i++ {
+	for i := 1; i < 1500; i++ {
 
 		// Add a change: Send an unsolicited doc revision in a rev request
 		receviedChangesWg.Add(1)

--- a/rest/blip_sync.go
+++ b/rest/blip_sync.go
@@ -24,7 +24,7 @@ import (
 // This connection remains open until the client closes it, and can receive any number of requests.
 type blipSyncContext struct {
 	blipContext        *blip.Context
-	sender             *blip.Sender
+	// sender             *blip.Sender
 	dbc                *db.DatabaseContext
 	user               auth.User
 	effectiveUsername  string
@@ -121,11 +121,18 @@ func (h *handler) handleBLIPSync() error {
 	return nil
 }
 
+
+
 // Registers a BLIP handler including the outer-level work of logging & error handling.
 // Includes the outer handler as a nested function.
 func (ctx *blipSyncContext) register(profile string, handlerFn func(*blipHandler, *blip.Message) error) {
+
 	ctx.blipContext.HandlerForProfile[profile] = func(rq *blip.Message) {
-		ctx.sender = rq.Sender
+
+		// ctx.sender = rq.Sender  // TODO: why isn't sender set for lifetime?  is this lazy loading?
+		// TODO: will requestsender change in scope of replication?
+		// TODO:
+
 		base.LogTo("Sync", "%s %q ... %s", rq, profile, ctx.effectiveUsername)
 
 		db, _ := db.GetDatabase(ctx.dbc, ctx.user)
@@ -234,12 +241,12 @@ func (bh *blipHandler) handleSubscribeToChanges(rq *blip.Message) error {
 	} else if filter != "" {
 		return base.HTTPErrorf(http.StatusBadRequest, "Unknown filter; try sync_gateway/bychannel")
 	}
-	go bh.sendChanges(since)
+	go bh.sendChanges(since, rq.Sender)  // TODO: does this ever end?
 	return nil
 }
 
 // Sends all changes since the given sequence
-func (bh *blipHandler) sendChanges(since db.SequenceID) {
+func (bh *blipHandler) sendChanges(since db.SequenceID, sender *blip.Sender) {
 	defer func() {
 		if panicked := recover(); panicked != nil {
 			base.Warn("*** PANIC sending changes: %v\n%s", panicked, debug.Stack())
@@ -265,7 +272,7 @@ func (bh *blipHandler) sendChanges(since db.SequenceID) {
 	pendingChanges := make([][]interface{}, 0, bh.batchSize)
 	sendPendingChangesAt := func(minChanges int) {
 		if len(pendingChanges) >= minChanges {
-			bh.sendBatchOfChanges(pendingChanges)
+			bh.sendBatchOfChanges(pendingChanges, sender)
 			pendingChanges = make([][]interface{}, 0, bh.batchSize)
 		}
 	}
@@ -288,24 +295,24 @@ func (bh *blipHandler) sendChanges(since db.SequenceID) {
 			sendPendingChangesAt(1)
 			if !caughtUp {
 				caughtUp = true
-				bh.sendBatchOfChanges(nil) // Signal to client that it's caught up
+				bh.sendBatchOfChanges(nil, sender) // Signal to client that it's caught up
 			}
 		}
 		return nil
 	})
 }
 
-func (bh *blipHandler) sendBatchOfChanges(changeArray [][]interface{}) {
+func (bh *blipHandler) sendBatchOfChanges(changeArray [][]interface{}, sender *blip.Sender) {
 	outrq := blip.NewRequest()
 	outrq.SetProfile("changes")
 	outrq.SetJSONBody(changeArray)
 	if len(changeArray) > 0 {
 		// Spawn a goroutine to await the client's response:
-		bh.sender.Send(outrq)
-		go bh.handleChangesResponse(outrq.Response(), changeArray)
+		sender.Send(outrq)
+		go bh.handleChangesResponse(outrq.Response(), changeArray, sender)
 	} else {
 		outrq.SetNoReply(true)
-		bh.sender.Send(outrq)
+		sender.Send(outrq)
 	}
 	if len(changeArray) > 0 {
 		base.LogTo("Sync", "Sent %d changes to client, from seq %v ... %s", len(changeArray), changeArray[0][0], bh.effectiveUsername)
@@ -315,7 +322,7 @@ func (bh *blipHandler) sendBatchOfChanges(changeArray [][]interface{}) {
 }
 
 // Handles the response to a pushed "changes" message, i.e. the list of revisions the client wants
-func (bh *blipHandler) handleChangesResponse(response *blip.Message, changeArray [][]interface{}) {
+func (bh *blipHandler) handleChangesResponse(response *blip.Message, changeArray [][]interface{}, sender *blip.Sender) {
 	defer func() {
 		if panicked := recover(); panicked != nil {
 			base.Warn("*** PANIC handling 'changes' response: %v\n%s", panicked, debug.Stack())
@@ -357,7 +364,7 @@ func (bh *blipHandler) handleChangesResponse(response *blip.Message, changeArray
 					return
 				}
 			}
-			bh.sendRevision(seq, docID, revID, knownRevs, maxHistory)
+			bh.sendRevision(seq, docID, revID, knownRevs, maxHistory, sender)
 		}
 	}
 }
@@ -446,7 +453,7 @@ func (bh *blipHandler) handleProposedChanges(rq *blip.Message) error {
 //////// DOCUMENTS:
 
 // Pushes a revision body to the client
-func (bh *blipHandler) sendRevision(seq db.SequenceID, docID string, revID string, knownRevs map[string]bool, maxHistory int) {
+func (bh *blipHandler) sendRevision(seq db.SequenceID, docID string, revID string, knownRevs map[string]bool, maxHistory int, sender *blip.Sender) {
 	base.LogTo("Sync+", "Sending rev %q %s based on %d known ... %s", docID, revID, len(knownRevs), bh.effectiveUsername)
 	body, err := bh.db.GetRev(docID, revID, true, nil)
 	if err != nil {
@@ -485,14 +492,14 @@ func (bh *blipHandler) sendRevision(seq db.SequenceID, docID string, revID strin
 	if atts := db.BodyAttachments(body); atts != nil {
 		// Allow client to download attachments in 'atts', but only while pulling this rev
 		bh.addAllowedAttachments(atts)
-		bh.sender.Send(outrq)
+		sender.Send(outrq)
 		go func() {
 			defer bh.removeAllowedAttachments(atts)
 			outrq.Response() // blocks till reply is received
 		}()
 	} else {
 		outrq.SetNoReply(true)
-		bh.sender.Send(outrq)
+		sender.Send(outrq)
 	}
 }
 
@@ -528,7 +535,7 @@ func (bh *blipHandler) handleAddRevision(rq *blip.Message) error {
 	}
 
 	// Check for any attachments I don't have yet, and request them:
-	if err := bh.downloadOrVerifyAttachments(body, minRevpos); err != nil {
+	if err := bh.downloadOrVerifyAttachments(body, minRevpos, rq.Sender); err != nil {
 		return err
 	}
 
@@ -560,7 +567,7 @@ func (bh *blipHandler) handleGetAttachment(rq *blip.Message) error {
 
 // For each attachment in the revision, makes sure it's in the database, asking the client to
 // upload it if necessary. This method blocks until all the attachments have been processed.
-func (bh *blipHandler) downloadOrVerifyAttachments(body db.Body, minRevpos int) error {
+func (bh *blipHandler) downloadOrVerifyAttachments(body db.Body, minRevpos int, sender *blip.Sender) error {
 	return bh.db.ForEachStubAttachment(body, minRevpos,
 		func(name string, digest string, knownData []byte, meta map[string]interface{}) ([]byte, error) {
 			if knownData != nil {
@@ -573,7 +580,7 @@ func (bh *blipHandler) downloadOrVerifyAttachments(body db.Body, minRevpos int) 
 				outrq := blip.NewRequest()
 				outrq.Properties = map[string]string{"Profile": "proveAttachment", "digest": digest}
 				outrq.SetBody(nonce)
-				bh.sender.Send(outrq)
+				sender.Send(outrq)
 				if body, err := outrq.Response().Body(); err != nil {
 					return nil, err
 				} else if string(body) != proof {
@@ -589,7 +596,7 @@ func (bh *blipHandler) downloadOrVerifyAttachments(body db.Body, minRevpos int) 
 				if isCompressible(name, meta) {
 					outrq.Properties["compress"] = "true"
 				}
-				bh.sender.Send(outrq)
+				sender.Send(outrq)
 				return outrq.Response().Body()
 			}
 		})

--- a/rest/blip_sync.go
+++ b/rest/blip_sync.go
@@ -24,7 +24,6 @@ import (
 // This connection remains open until the client closes it, and can receive any number of requests.
 type blipSyncContext struct {
 	blipContext        *blip.Context
-	// sender             *blip.Sender
 	dbc                *db.DatabaseContext
 	user               auth.User
 	effectiveUsername  string
@@ -128,10 +127,6 @@ func (h *handler) handleBLIPSync() error {
 func (ctx *blipSyncContext) register(profile string, handlerFn func(*blipHandler, *blip.Message) error) {
 
 	ctx.blipContext.HandlerForProfile[profile] = func(rq *blip.Message) {
-
-		// ctx.sender = rq.Sender  // TODO: why isn't sender set for lifetime?  is this lazy loading?
-		// TODO: will requestsender change in scope of replication?
-		// TODO:
 
 		base.LogTo("Sync", "%s %q ... %s", rq, profile, ctx.effectiveUsername)
 
@@ -241,12 +236,12 @@ func (bh *blipHandler) handleSubscribeToChanges(rq *blip.Message) error {
 	} else if filter != "" {
 		return base.HTTPErrorf(http.StatusBadRequest, "Unknown filter; try sync_gateway/bychannel")
 	}
-	go bh.sendChanges(since, rq.Sender)  // TODO: does this ever end?
+	go bh.sendChanges(rq.Sender, since)  // TODO: does this ever end?
 	return nil
 }
 
 // Sends all changes since the given sequence
-func (bh *blipHandler) sendChanges(since db.SequenceID, sender *blip.Sender) {
+func (bh *blipHandler) sendChanges(sender *blip.Sender, since db.SequenceID) {
 	defer func() {
 		if panicked := recover(); panicked != nil {
 			base.Warn("*** PANIC sending changes: %v\n%s", panicked, debug.Stack())
@@ -272,7 +267,7 @@ func (bh *blipHandler) sendChanges(since db.SequenceID, sender *blip.Sender) {
 	pendingChanges := make([][]interface{}, 0, bh.batchSize)
 	sendPendingChangesAt := func(minChanges int) {
 		if len(pendingChanges) >= minChanges {
-			bh.sendBatchOfChanges(pendingChanges, sender)
+			bh.sendBatchOfChanges(sender, pendingChanges)
 			pendingChanges = make([][]interface{}, 0, bh.batchSize)
 		}
 	}
@@ -295,21 +290,21 @@ func (bh *blipHandler) sendChanges(since db.SequenceID, sender *blip.Sender) {
 			sendPendingChangesAt(1)
 			if !caughtUp {
 				caughtUp = true
-				bh.sendBatchOfChanges(nil, sender) // Signal to client that it's caught up
+				bh.sendBatchOfChanges(sender,nil) // Signal to client that it's caught up
 			}
 		}
 		return nil
 	})
 }
 
-func (bh *blipHandler) sendBatchOfChanges(changeArray [][]interface{}, sender *blip.Sender) {
+func (bh *blipHandler) sendBatchOfChanges(sender *blip.Sender, changeArray [][]interface{}) {
 	outrq := blip.NewRequest()
 	outrq.SetProfile("changes")
 	outrq.SetJSONBody(changeArray)
 	if len(changeArray) > 0 {
 		// Spawn a goroutine to await the client's response:
 		sender.Send(outrq)
-		go bh.handleChangesResponse(outrq.Response(), changeArray, sender)
+		go bh.handleChangesResponse(sender, outrq.Response(), changeArray)
 	} else {
 		outrq.SetNoReply(true)
 		sender.Send(outrq)
@@ -322,7 +317,7 @@ func (bh *blipHandler) sendBatchOfChanges(changeArray [][]interface{}, sender *b
 }
 
 // Handles the response to a pushed "changes" message, i.e. the list of revisions the client wants
-func (bh *blipHandler) handleChangesResponse(response *blip.Message, changeArray [][]interface{}, sender *blip.Sender) {
+func (bh *blipHandler) handleChangesResponse(sender *blip.Sender, response *blip.Message, changeArray [][]interface{}) {
 	defer func() {
 		if panicked := recover(); panicked != nil {
 			base.Warn("*** PANIC handling 'changes' response: %v\n%s", panicked, debug.Stack())
@@ -364,7 +359,7 @@ func (bh *blipHandler) handleChangesResponse(response *blip.Message, changeArray
 					return
 				}
 			}
-			bh.sendRevision(seq, docID, revID, knownRevs, maxHistory, sender)
+			bh.sendRevision(sender, seq, docID, revID, knownRevs, maxHistory)
 		}
 	}
 }
@@ -453,7 +448,7 @@ func (bh *blipHandler) handleProposedChanges(rq *blip.Message) error {
 //////// DOCUMENTS:
 
 // Pushes a revision body to the client
-func (bh *blipHandler) sendRevision(seq db.SequenceID, docID string, revID string, knownRevs map[string]bool, maxHistory int, sender *blip.Sender) {
+func (bh *blipHandler) sendRevision(sender *blip.Sender, seq db.SequenceID, docID string, revID string, knownRevs map[string]bool, maxHistory int) {
 	base.LogTo("Sync+", "Sending rev %q %s based on %d known ... %s", docID, revID, len(knownRevs), bh.effectiveUsername)
 	body, err := bh.db.GetRev(docID, revID, true, nil)
 	if err != nil {

